firebase-firestore/ktx/src/main/kotlin/com/google/firebase/firestore/ktx/serialization/NestedMapEncoder.kt

```diff
@@ -0,0 +1,155 @@
+// Copyright 2022 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package com.google.firebase.firestore.ktx.serialization
+
+import kotlinx.serialization.SerializationStrategy
+import kotlinx.serialization.descriptors.SerialDescriptor
+import kotlinx.serialization.descriptors.StructureKind
+import kotlinx.serialization.descriptors.elementNames
+import kotlinx.serialization.encoding.AbstractEncoder
+import kotlinx.serialization.encoding.CompositeEncoder
+import kotlinx.serialization.modules.EmptySerializersModule
+import kotlinx.serialization.modules.SerializersModule
+import kotlinx.serialization.serializer
+
+class NestedMapEncoder(
+    private var map: MutableMap<Int, MutableMap<String, Any?>> = mutableMapOf(),
+    private var depth: Int = 0,
+    private var list: MutableList<Any> = mutableListOf(),
+    private var descriptor: SerialDescriptor? = null
+) : AbstractEncoder() {
+
+    companion object {
+        private const val ROOT_LEVEL: Int = 1
+    }
+
+    private var elementIndex: Int = 0
+
+    init {
+        if (depth == ROOT_LEVEL) map[ROOT_LEVEL] = mutableMapOf()
+    }
+
+    fun serializedResult() = map[ROOT_LEVEL]!!
+
+    override fun encodeEnum(enumDescriptor: SerialDescriptor, index: Int) =
+        enumDescriptor.elementNames.toList().let { encodeValue(it.get(index)) }
+
+    override val serializersModule: SerializersModule = EmptySerializersModule
+
+    override fun encodeNull() {
+        val key: String = list[elementIndex++] as String
```

Comment:
```suggestion
        val key = list[elementIndex++] as String
```

Reply:
Here and below

Reply:
Thank you, I have removed all the `: Type`  for the cast statements in this file.